Set conditions properly while deploying.

awels opened the pull request:
Signed-off-by: Alexander Wels <awels@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
The conditions were not properly being set during deployment (Available too soon). This fixes the conditions to match the following:
Start deploy:
Available: false
Progression: true
Degraded: false

During deploy stay on that.
Once deploy is finished and all pods are available:
Available: true
Progression: false
Degraded: false

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

mhenriks commented on pkg/operator/controller/controller_test.go:
```diff
@@ -216,69 +216,9 @@ var _ = Describe("Controller", func() {
-					}
-				}
-
-				Expect(running).To(BeTrue())
```
lot of red here.  not sure how this is all useless/invalid now.

awels replied:
That is a duplicate of setDeploymentsReady(args), the code is identical. So instead of duplicating it, I just replaced it.

mhenriks replied:
part of it is identical.  not the whole thing.

awels replied:
Actually you are right, first it setting the Status.Replicas to a valid number, and then it checks what the conditions are, then it sets the ReadyReplicas and checks the conditions again. Where setDeploymentsReady sets both at once and then checks the conditions. If we are interested in the intermediate check, I can put it back.

mhenriks commented:
>Start deploy:
Available: false
Progression: true
Degraded: false

Why not degraded here as well?  Under what conditions are we available and degraded?  And how is that different than not available and not degraded?

mhenriks commented on pkg/operator/controller/controller_test.go:
```diff
@@ -1314,6 +1264,29 @@ func setDeploymentsReady(args *args) bool {
+		}
+
+	}
+	doReconcile(args)
```
should conditions be verified here?

awels replied:
I am assuming you are asking if I should verify if the degraded condition is true. The problem is depending on the state of the CR, degraded can be either true or false (if I am deploying degraded is always false, but if I am deployed already, then degraded will be true). I guess I can check if I am deploying or not and check based on that.

awels commented:
https://github.com/kubevirt/hyperconverged-cluster-operator/blob/master/docs/conditions.md has all the combinations of conditions. Deploying is not degraded. But we can be degraded and available after we have been deployed and a lets say one of our deployment pods goes down, while its restarting the pod. If for instance the upload proxy goes down most of CDI is available except upload. So available and degraded.

mhenriks commented:
> https://github.com/kubevirt/hyperconverged-cluster-operator/blob/master/docs/conditions.md has all the combinations of conditions. Deploying is not degraded. But we can be degraded and available after we have been deployed and a lets say one of our deployment pods goes down, while its restarting the pod. If for instance the upload proxy goes down most of CDI is available except upload. So available and degraded.

Okay, according to the doc, there are 6 possible condition value combinations.  Do we support all of them?  And if so, do we have tests for each?

awels commented:
> > https://github.com/kubevirt/hyperconverged-cluster-operator/blob/master/docs/conditions.md has all the combinations of conditions. Deploying is not degraded. But we can be degraded and available after we have been deployed and a lets say one of our deployment pods goes down, while its restarting the pod. If for instance the upload proxy goes down most of CDI is available except upload. So available and degraded.
> 
> Okay, according to the doc, there are 6 possible condition value combinations. Do we support all of them? And if so, do we have tests for each?

I put in helper functions to each combination (cr.go), and one of them was never called so I removed it. So technically we don't 'support' all 6 since we never encounter one combination.

As for the tests, the combinations mapped relatively easily onto our existing conditions, so I just modified those unit tests to check for the new conditions.
